File: S2/peerlearn/app/server/API/Controllers/Auth/OtpController.cs
using API.Contracts.Auth;
using API.Models;
using API.Models.Auth;
using Core.Interfaces.Services;
using Microsoft.AspNetCore.Mvc;

namespace API.Controllers.Auth;

[ApiController]
[Route("api/[controller]")]
public class OtpController : ControllerBase {
    private readonly IOtpService _otpService;

    public OtpController(IOtpService otpService) {
        _otpService = otpService;
    }

    [HttpPost("send")]
    public async Task<IActionResult> SendOtp([FromBody] SendOtpRequest? request) {
        if (request == null || string.IsNullOrWhiteSpace(request.Email)) {
            return BadRequest(new ApiResponse<EmptyResponse> {
                Success = false,
                Message = "Invalid request",
                Data = null
            });
        }
        DateTime expirationTime = await _otpService.SendOtpAsync(request.Email);
        return Ok(new ApiResponse<OtpResponse> {
            Success = true,
            Message = "OTP sent successfully",
            Data = new OtpResponse { ExpiresAt = expirationTime }
        });
    }

    [HttpPut("verify")]
    public async Task<IActionResult> VerifyOtp([FromBody] OtpVerifyRequest? request) {
        if (request == null || string.IsNullOrWhiteSpace(request.Email) || string.IsNullOrWhiteSpace(request.Code)) {
            return BadRequest(new ApiResponse<EmptyResponse> {
                Success = false,
                Message = "Invalid request",
                Data = null
            });
        }
        await _otpService.VerifyOtpAsync(request.Email, request.Code);
        return Ok(new ApiResponse<EmptyResponse> {
            Success = true,
            Message = "OTP verified successfully",
            Data = null
        });
    }
}

File: S2/ticktick-app/server/src/controllers/admin/otpController.ts
import { NextFunction, Response, Request } from "express";
import { OtpService } from "../../services/admin/otpService";

export class OtpController {

    private otpService = new OtpService();

    public async sendOtp(req: Request, res: Response, next: NextFunction){
        const { email } = req.body;
    
        try{
          await this.otpService.sendOtp(email);
          res.status(200).json({
            success: true,
            message: `OTP send successfully to ${email}`
          });
        } catch(err){
          next(err);
        }
      }
    
      public async verifyOtp(req: Request, res: Response, next: NextFunction){
        const { email, otp } = req.body
    
        try{
          const isOtpValid = await this.otpService.validateOtp(email, otp);
    
          if(isOtpValid){
            res.status(200).json({
              success: true,
              message: "OTP verified successfully"
            })
          }
        } catch(err){
          next(err)
        }
      }
}